File: regions/system/library/csharp/graphql-client/tests/K1s0.GraphQlClient.Tests/GraphQlClientTests.cs
using K1s0.GraphQlClient;

namespace K1s0.GraphQlClient.Tests;

public class GraphQlClientTests
{
    [Fact]
    public void GraphQlQuery_CreatesWithRequiredFields()
    {
        var query = new GraphQlQuery("{ users { id } }");
        Assert.Equal("{ users { id } }", query.Query);
        Assert.Null(query.Variables);
        Assert.Null(query.OperationName);
    }

    [Fact]
    public void GraphQlQuery_CreatesWithAllFields()
    {
        var vars = new Dictionary<string, object> { ["id"] = "1" };
        var query = new GraphQlQuery("query GetUser($id: ID!)", vars, "GetUser");
        Assert.Equal("GetUser", query.OperationName);
        Assert.NotNull(query.Variables);
    }

    [Fact]
    public void GraphQlResponse_HasErrors_FalseWhenNoErrors()
    {
        var resp = new GraphQlResponse<string>("ok");
        Assert.False(resp.HasErrors);
    }

    [Fact]
    public void GraphQlResponse_HasErrors_TrueWithErrors()
    {
        var resp = new GraphQlResponse<string>(null, new List<GraphQlError> { new("fail") });
        Assert.True(resp.HasErrors);
    }

    [Fact]
    public void GraphQlResponse_HasErrors_FalseWithEmptyList()
    {
        var resp = new GraphQlResponse<string>(null, new List<GraphQlError>());
        Assert.False(resp.HasErrors);
    }

    [Fact]
    public void ErrorLocation_StoresLineAndColumn()
    {
        var loc = new ErrorLocation(3, 5);
        Assert.Equal(3, loc.Line);
        Assert.Equal(5, loc.Column);
    }

    [Fact]
    public void GraphQlError_CreatesWithMessage()
    {
        var err = new GraphQlError("Not found");
        Assert.Equal("Not found", err.Message);
        Assert.Null(err.Locations);
        Assert.Null(err.Path);
    }

    [Fact]
    public async Task ExecuteAsync_ReturnsConfiguredResponse()
    {
        var client = new InMemoryGraphQlClient();
        client.SetResponse("GetUser", "Alice");

        var result = await client.ExecuteAsync<string>(
            new GraphQlQuery("query GetUser { user }", OperationName: "GetUser"));

        Assert.False(result.HasErrors);
        Assert.Equal("Alice", result.Data);
    }

    [Fact]
    public async Task ExecuteAsync_ReturnsErrorForUnconfigured()
    {
        var client = new InMemoryGraphQlClient();

        var result = await client.ExecuteAsync<string>(
            new GraphQlQuery("{ unknown }", OperationName: "Unknown"));

        Assert.True(result.HasErrors);
        Assert.Null(result.Data);
    }

    [Fact]
    public async Task ExecuteMutationAsync_ReturnsConfiguredResponse()
    {
        var client = new InMemoryGraphQlClient();
        client.SetResponse("CreateUser", "Bob");

        var result = await client.ExecuteMutationAsync<string>(
            new GraphQlQuery("mutation CreateUser { createUser }", OperationName: "CreateUser"));

        Assert.False(result.HasErrors);
        Assert.Equal("Bob", result.Data);
    }

    [Fact]
    public async Task ExecuteAsync_FallsBackToQueryText()
    {
        var client = new InMemoryGraphQlClient();
        client.SetResponse("{ me { id } }", "me-data");

        var result = await client.ExecuteAsync<string>(
            new GraphQlQuery("{ me { id } }"));

        Assert.False(result.HasErrors);
        Assert.Equal("me-data", result.Data);
    }

    [Fact]
    public async Task ExecuteAsync_TypeMismatchReturnsError()
    {
        var client = new InMemoryGraphQlClient();
        client.SetResponse("Op", 42);

        var result = await client.ExecuteAsync<string>(
            new GraphQlQuery("query", OperationName: "Op"));

        Assert.True(result.HasErrors);
    }
}

File: regions/system/library/swift/graphql-client/Tests/K1s0GraphQlClientTests/GraphQlClientTests.swift
import Testing
@testable import K1s0GraphQlClient

@Suite("GraphQlClient Tests")
struct GraphQlClientTests {
    @Test("クエリを実行してレスポンスを取得できること")
    func testExecute() async throws {
        let client = InMemoryGraphQlClient()
        let response = GraphQlResponse<[String: String]>(data: ["name": "test"])
        await client.setResponse(response, forOperation: "GetUser")

        let query = GraphQlQuery(query: "{ user { name } }", operationName: "GetUser")
        let result: GraphQlResponse<[String: String]> = try await client.execute(query: query)
        #expect(result.data?["name"] == "test")
        #expect(result.hasErrors == false)
    }

    @Test("ミューテーションを実行してレスポンスを取得できること")
    func testExecuteMutation() async throws {
        let client = InMemoryGraphQlClient()
        let response = GraphQlResponse<[String: String]>(data: ["id": "123"])
        await client.setResponse(response, forOperation: "CreateUser")

        let mutation = GraphQlQuery(query: "mutation { createUser { id } }", operationName: "CreateUser")
        let result: GraphQlResponse<[String: String]> = try await client.executeMutation(mutation: mutation)
        #expect(result.data?["id"] == "123")
    }

    @Test("存在しないオペレーションでエラーになること")
    func testOperationNotFound() async throws {
        let client = InMemoryGraphQlClient()
        let query = GraphQlQuery(query: "{ user { name } }", operationName: "Unknown")

        do {
            let _: GraphQlResponse<[String: String]> = try await client.execute(query: query)
            #expect(Bool(false), "Should have thrown")
        } catch is GraphQlClientError {
            // expected
        }
    }

    @Test("オペレーション名なしでエラーになること")
    func testUnknownOperation() async throws {
        let client = InMemoryGraphQlClient()
        let query = GraphQlQuery(query: "{ user { name } }")

        do {
            let _: GraphQlResponse<[String: String]> = try await client.execute(query: query)
            #expect(Bool(false), "Should have thrown")
        } catch is GraphQlClientError {
            // expected
        }
    }

    @Test("エラーレスポンスを返すこと")
    func testErrorResponse() async throws {
        let client = InMemoryGraphQlClient()
        let error = GraphQlError(message: "Not found", locations: [ErrorLocation(line: 1, column: 5)], path: ["user"])
        let response = GraphQlResponse<[String: String]>(data: nil, errors: [error])
        await client.setResponse(response, forOperation: "GetUser")

        let query = GraphQlQuery(query: "{ user { name } }", operationName: "GetUser")
        let result: GraphQlResponse<[String: String]> = try await client.execute(query: query)
        #expect(result.hasErrors == true)
        #expect(result.errors?.first?.message == "Not found")
        #expect(result.errors?.first?.locations?.first?.line == 1)
        #expect(result.errors?.first?.path?.first == "user")
    }

    @Test("GraphQlQueryの変数が設定できること")
    func testQueryWithVariables() {
        let query = GraphQlQuery(
            query: "query($id: ID!) { user(id: $id) { name } }",
            variables: ["id": "123"],
            operationName: "GetUser"
        )
        #expect(query.query.contains("$id"))
        #expect(query.operationName == "GetUser")
    }

    @Test("GraphQlClientErrorの各バリアント")
    func testErrorVariants() {
        let err1 = GraphQlClientError.operationNotFound(name: "Test")
        if case .operationNotFound(let name) = err1 {
            #expect(name == "Test")
        }

        let err2 = GraphQlClientError.typeMismatch
        if case .typeMismatch = err2 {
            // expected
        } else {
            #expect(Bool(false), "Should be typeMismatch")
        }

        let err3 = GraphQlClientError.unknownOperation
        if case .unknownOperation = err3 {
            // expected
        } else {
            #expect(Bool(false), "Should be unknownOperation")
        }
    }
}
